File: python/src/resp_bench/client/impl/recording_client.py
"""In-memory recording client for server-free testing.

Records operations and supports simulated latency and error injection via
``specific_driver_config`` (``operation_delay_micros``,
``delay_variation_micros``, ``error_rate``, ``error_message``). This lets the
integration tests exercise the full engine without a live server, mirroring the
Ruby recording driver's role.
"""

from __future__ import annotations

import asyncio
import random
import time
from dataclasses import dataclass
from typing import List, Optional

from ...config.driver_config import DriverConfig
from ..benchmark_client import AsyncBenchmarkClient
from ..timed_result import TimedResult


@dataclass
class RecordedOperation:
    command: str
    key: Optional[str]
    value: Optional[bytes]
    success: bool
    error_message: Optional[str]


class RecordingClient(AsyncBenchmarkClient):
    def __init__(self) -> None:
        self.operations: List[RecordedOperation] = []
        self._stored_data: dict = {}
        self._connected = False
        self._operation_delay_micros = 0
        self._delay_variation_micros = 0
        self._error_rate = 0.0
        self._error_message = "Simulated error"
        self._random = random.Random()

    async def connect(self, host: str, port: int, config: DriverConfig) -> None:
        self._connected = True
        cfg = config.specific_driver_config or {}
        self._operation_delay_micros = int(cfg.get("operation_delay_micros", 0) or 0)
        self._delay_variation_micros = int(cfg.get("delay_variation_micros", 0) or 0)
        self._error_rate = float(cfg.get("error_rate", 0.0) or 0.0)
        self._error_message = str(cfg.get("error_message", "Simulated error"))
        self.operations.append(RecordedOperation("CONNECT", None, None, True, None))

    async def ping(self) -> TimedResult:
        latency, success = await self._simulate()
        error = None if success else self._error_message
        self.operations.append(RecordedOperation("PING", None, None, success, error))
        if success:
            return TimedResult(value="PONG", latency_micros=latency)
        return TimedResult(value=None, latency_micros=latency, error=RuntimeError(error))

    async def get(self, key: str) -> TimedResult:
        latency, success = await self._simulate()
        error = None if success else self._error_message
        self.operations.append(RecordedOperation("GET", key, None, success, error))
        if success:
            return TimedResult(value=self._stored_data.get(key), latency_micros=latency)
        return TimedResult(value=None, latency_micros=latency, error=RuntimeError(error))

    async def set(self, key: str, value: bytes) -> TimedResult:
        latency, success = await self._simulate()
        error = None if success else self._error_message
        if success:
            self._stored_data[key] = value
        self.operations.append(RecordedOperation("SET", key, value, success, error))
        if success:
            return TimedResult(value="OK", latency_micros=latency)
        return TimedResult(value=None, latency_micros=latency, error=RuntimeError(error))

    async def close(self) -> None:
        self._connected = False
        self.operations.append(RecordedOperation("CLOSE", None, None, True, None))

    def driver_version(self) -> str:
        return "1.0.0"

    async def _simulate(self) -> tuple[int, bool]:
        start = time.perf_counter_ns()
        delay_micros = self._calculate_delay_micros()
        if delay_micros > 0:
            await asyncio.sleep(delay_micros / 1_000_000)
        latency = (time.perf_counter_ns() - start) // 1000
        return latency, not self._should_simulate_error()

    def _calculate_delay_micros(self) -> int:
        if self._operation_delay_micros <= 0:
            return 0
        delay = self._operation_delay_micros
        if self._delay_variation_micros > 0:
            variation = self._random.randint(
                -self._delay_variation_micros, self._delay_variation_micros
            )
            delay = max(0, delay + variation)
        return delay

    def _should_simulate_error(self) -> bool:
        if self._error_rate <= 0.0:
            return False
        if self._error_rate >= 1.0:
            return True
        return self._random.random() < self._error_rate

File: node/src/client/impl/recordingClient.ts
/**
 * In-memory recording client for server-free testing.
 *
 * Records operations and supports simulated latency and error injection via
 * `specific_driver_config` (`operation_delay_micros`, `delay_variation_micros`,
 * `error_rate`, `error_message`). This lets the integration tests exercise the
 * whole engine without a live server, mirroring the Ruby/Python recording
 * drivers.
 */

import type { DriverConfig } from '../../config/driverConfig.js';
import type { BenchmarkClient } from '../benchmarkClient.js';
import type { TimedResult } from '../timedResult.js';

export interface RecordedOperation {
  command: string;
  key: string | null;
  value: Buffer | null;
  success: boolean;
  errorMessage: string | null;
}

const NANOS_PER_MICRO = 1000n;

function readNumber(source: Record<string, unknown>, key: string, fallback: number): number {
  const value = source[key];
  return typeof value === 'number' && Number.isFinite(value) ? value : fallback;
}

/** Sleep for whole microseconds, tolerating setTimeout's ~1ms floor. */
function sleepMicros(micros: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, micros / 1000));
}

export class RecordingClient implements BenchmarkClient {
  readonly operations: RecordedOperation[] = [];
  private readonly storedData = new Map<string, Buffer>();
  private operationDelayMicros = 0;
  private delayVariationMicros = 0;
  private errorRate = 0;
  private errorMessage = 'Simulated error';
  private warmupMode = false;

  async connect(_host: string, _port: number, config: DriverConfig): Promise<void> {
    const cfg = config.specificDriverConfig;
    this.operationDelayMicros = readNumber(cfg, 'operation_delay_micros', 0);
    this.delayVariationMicros = readNumber(cfg, 'delay_variation_micros', 0);
    this.errorRate = readNumber(cfg, 'error_rate', 0);
    const message = cfg['error_message'];
    if (typeof message === 'string') this.errorMessage = message;
    this.record('CONNECT', null, null, true, null);
  }

  async ping(): Promise<TimedResult<string>> {
    const { latencyMicros, success } = await this.simulate();
    this.record('PING', null, null, success, success ? null : this.errorMessage);
    return success
      ? { value: 'PONG', latencyMicros }
      : { value: null, latencyMicros, error: new Error(this.errorMessage) };
  }

  async get(key: string): Promise<TimedResult<string>> {
    const { latencyMicros, success } = await this.simulate();
    this.record('GET', key, null, success, success ? null : this.errorMessage);
    if (!success) return { value: null, latencyMicros, error: new Error(this.errorMessage) };
    const stored = this.storedData.get(key);
    return { value: stored === undefined ? null : stored.toString('latin1'), latencyMicros };
  }

  async set(key: string, value: Buffer): Promise<TimedResult<string>> {
    const { latencyMicros, success } = await this.simulate();
    if (success) this.storedData.set(key, value);
    this.record('SET', key, value, success, success ? null : this.errorMessage);
    return success
      ? { value: 'OK', latencyMicros }
      : { value: null, latencyMicros, error: new Error(this.errorMessage) };
  }

  async close(): Promise<void> {
    this.record('CLOSE', null, null, true, null);
  }

  driverVersion(): string {
    return '1.0.0';
  }

  /**
   * Suppress simulated errors during warmup, matching Java's recording client.
   * The engine's warmup fail-fast exists to catch an unreachable server, so
   * injected errors must not trip it.
   */
  setWarmupMode(warmup: boolean): void {
    this.warmupMode = warmup;
  }

  private record(
    command: string,
    key: string | null,
    value: Buffer | null,
    success: boolean,
    errorMessage: string | null,
  ): void {
    this.operations.push({ command, key, value, success, errorMessage });
  }

  private async simulate(): Promise<{ latencyMicros: number; success: boolean }> {
    const start = process.hrtime.bigint();
    const delayMicros = this.calculateDelayMicros();
    if (delayMicros > 0) await sleepMicros(delayMicros);
    const latencyMicros = Number((process.hrtime.bigint() - start) / NANOS_PER_MICRO);
    return { latencyMicros, success: !this.shouldSimulateError() };
  }

  private calculateDelayMicros(): number {
    if (this.operationDelayMicros <= 0) return 0;
    let delay = this.operationDelayMicros;
    if (this.delayVariationMicros > 0) {
      const spread = 2 * this.delayVariationMicros + 1;
      const variation = Math.floor(Math.random() * spread) - this.delayVariationMicros;
      delay = Math.max(0, delay + variation);
    }
    return delay;
  }

  private shouldSimulateError(): boolean {
    if (this.warmupMode) return false;
    if (this.errorRate <= 0) return false;
    if (this.errorRate >= 1) return true;
    return Math.random() < this.errorRate;
  }
}
